### PR TITLE
Add ability to specify path to TF binary

### PR DESCRIPTION
**Pull request by paultyng**

### CODE REVIEW

File: langserver/langserver.go [hunk 1]
```diff
@@ -12,12 +12,15 @@ import (
...
 	"github.com/creachadair/jrpc2/server"
+
+	"github.com/hashicorp/terraform-ls/internal/terraform/discovery"
 	"github.com/hashicorp/terraform-ls/langserver/session"
 )
 
 type langServer struct {
 	srvCtx     context.Context
 	logger     *log.Logger
+	tfPath     discovery.DiscoveryFunc
```
Comment (radeksimko) [inline langserver/langserver.go:23]: Nitpick, but do you mind calling this `tfDiscoFunc` or something similar that makes it more obvious that this variable is a function and not a string?

I suppose the other `tfPath` in `service.go` is a forgotten relic that should also be changed - especially once https://github.com/hashicorp/terraform-ls/issues/69 is addressed. I'm happy to address it as part of #69 though - no biggie.

File: langserver/session/types.go [hunk 2]
```diff
@@ -5,12 +5,15 @@ import (
...
 	"github.com/creachadair/jrpc2"
+
+	"github.com/hashicorp/terraform-ls/internal/terraform/discovery"
 )
 
 type Session interface {
 	Assigner() (jrpc2.Assigner, error)
 	Finish(jrpc2.ServerStatus)
 	SetLogger(*log.Logger)
+	SetDiscovery(discovery.DiscoveryFunc)
```
Comment (radeksimko) [inline langserver/session/types.go:16]: Nitpick: Can we call this `SetDiscoveryFunc`, so it's more obvious that the argument is a function?

File: langserver/handlers/service.go [hunk 3]
```diff
@@ -53,6 +53,10 @@ func (svc *service) SetLogger(logger *log.Logger) {
 	svc.logger = logger
 }
 
+func (svc *service) SetDiscovery(f discovery.DiscoveryFunc) {
```
Comment (radeksimko) [inline langserver/handlers/service.go:56]: Nitpick: Can we call this `SetDiscoveryFunc`?

**Comment by paultyng** [comment]
@radeksimko made the suggested adjustments, also added some testing of the path string to make sure its a bit more valid as discussed. I cleaned up a few of the other `ui.Error` calls in another commit as well just for consistency. Let me know if its still good to merge.

File: commands/serve_command.go [hunk 4]
```diff
@@ -79,18 +82,43 @@ func (c *ServeCommand) Run(args []string) int {
...
+		}
+		stat, err := os.Stat(path)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Unable to find Terraform binary: %s", err))
+			return 1
+		}
+		if stat.IsDir() {
+			c.Ui.Error(fmt.Sprintf("Expected a Terraform binary, got a directory: %q", path))
+			return 1
+		}
```
Comment (radeksimko) [inline commands/serve_command.go:103]: Total nitpick, but I assume the above lines could make a `validateTfExecPath(string) error` function.

**Comment by paultyng** [comment]
The confusing part was they weren't on all errors just some errors, but if its the how it should be presented you could already create a fancy UI implementation that adds the separation.

I was hoping to maybe revisit/refactor this after the release anyway, clean up the hard coded exit codes and flag validation a bit. I'll just write an issue to do so if thats ok with you?

**Pull request merged**